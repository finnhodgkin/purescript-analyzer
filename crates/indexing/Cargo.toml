[package]
name = "indexing"
version = "0.1.0"
edition = "2021"

[dependencies]
fxhash = "0.2.1"
indexmap = "2.7.0"
la-arena = "0.3.1"
petgraph = "0.7.0"
rowan = "0.16.1"
rustc-hash = "2.1.0"
smol_str = "0.3.2"
syntax = { version = "0.1.0", path = "../syntax" }

[dev-dependencies]
lexing = { path = "../lexing" }
parsing = { path = "../parsing" }
