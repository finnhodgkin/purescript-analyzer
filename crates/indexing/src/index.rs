use std::sync::Arc;

use la_arena::{Arena, ArenaMap, Idx};
use rustc_hash::FxHashMap;
use smol_str::SmolStr;

use crate::source::*;

#[derive(Debug, PartialEq, Eq)]
pub enum TermItem {
    ClassMember { id: ClassMemberId },
    Constructor { id: DataConstructorId },
    Derive { id: DeriveId },
    Foreign { id: ForeignValueId },
    Instance { id: InstanceId },
    Operator { id: InfixId },
    Value { signature: Option<ValueSignatureId>, equations: Vec<ValueEquationId> },
}

pub type TermItemId = Idx<TermItem>;

#[derive(Debug, PartialEq, Eq)]
pub enum TypeItem {
    Data {
        signature: Option<DataSignatureId>,
        equation: Option<DataEquationId>,
        role: Option<TypeRoleId>,
    },
    Newtype {
        signature: Option<NewtypeSignatureId>,
        equation: Option<NewtypeEquationId>,
        role: Option<TypeRoleId>,
    },
    Synonym {
        signature: Option<TypeSignatureId>,
        equation: Option<TypeEquationId>,
    },
    Class {
        signature: Option<ClassSignatureId>,
        declaration: Option<ClassDeclarationId>,
    },
    Foreign {
        id: ForeignDataId,
    },
    Operator {
        id: InfixId,
    },
}

pub type TypeItemId = Idx<TypeItem>;

#[derive(Debug, PartialEq, Eq)]
pub enum ImplicitItems {
    Everything,
    Enumerated(Arc<[SmolStr]>),
}

pub type ImportedTerms = FxHashMap<SmolStr, ImportItemId>;
pub type ImportedTypes = FxHashMap<SmolStr, (ImportItemId, Option<ImplicitItems>)>;

#[derive(Debug, PartialEq, Eq)]
pub struct ImportedItems {
    pub terms: ImportedTerms,
    pub types: ImportedTypes,
    pub exported: bool,
}

#[derive(Debug, Default, PartialEq, Eq)]
pub struct Index {
    term_item: Arena<TermItem>,
    type_item: Arena<TypeItem>,
    term_export: ArenaMap<TermItemId, ExportItemId>,
    type_export: ArenaMap<TypeItemId, ExportItemId>,
    alias_nominal: FxHashMap<SmolStr, Vec<ImportId>>,
    imported_items: FxHashMap<ImportId, ImportedItems>,
    term_nominal: FxHashMap<SmolStr, TermItemId>,
    type_nominal: FxHashMap<SmolStr, TypeItemId>,
}

impl std::ops::Index<TermItemId> for Index {
    type Output = TermItem;

    fn index(&self, id: TermItemId) -> &TermItem {
        &self.term_item[id]
    }
}

impl std::ops::Index<TypeItemId> for Index {
    type Output = TypeItem;

    fn index(&self, id: TypeItemId) -> &TypeItem {
        &self.type_item[id]
    }
}

#[derive(Debug, Default, PartialEq, Eq)]
pub struct Relational {
    data: Vec<(TypeItemId, TermItemId)>,
    class: Vec<(TypeItemId, TermItemId)>,
    chain: Vec<(InstanceChainId, InstanceId)>,
    instance: Vec<(InstanceId, InstanceMemberId)>,
}

impl Index {
    pub(crate) fn insert_term_item(&mut self, k: Option<SmolStr>, v: TermItem) -> TermItemId {
        let id = self.term_item.alloc(v);
        let Some(k) = k else { return id };
        self.term_nominal.insert(k, id);
        id
    }

    pub(crate) fn term_item_mut(&mut self, k: &str) -> Option<(&mut TermItem, TermItemId)> {
        let &id = self.term_nominal.get(k)?;
        Some((&mut self.term_item[id], id))
    }

    pub(crate) fn export_term_item(&mut self, item_id: TermItemId, id: ExportItemId) {
        self.term_export.insert(item_id, id);
    }

    pub(crate) fn term_item_export(&self, id: TermItemId) -> Option<ExportItemId> {
        self.term_export.get(id).copied()
    }

    pub(crate) fn insert_type_item(&mut self, k: Option<SmolStr>, v: TypeItem) -> TypeItemId {
        let id = self.type_item.alloc(v);
        let Some(k) = k else { return id };
        self.type_nominal.insert(k, id);
        id
    }

    pub(crate) fn type_item_mut(&mut self, k: &str) -> Option<(&mut TypeItem, TypeItemId)> {
        let &id = self.type_nominal.get(k)?;
        Some((&mut self.type_item[id], id))
    }

    pub(crate) fn export_type_item(&mut self, item_id: TypeItemId, id: ExportItemId) {
        self.type_export.insert(item_id, id);
    }

    pub(crate) fn type_item_export(&self, id: TypeItemId) -> Option<ExportItemId> {
        self.type_export.get(id).copied()
    }

    pub(crate) fn insert_import_alias(&mut self, k: SmolStr, v: ImportId) {
        self.alias_nominal.entry(k).or_default().push(v);
    }

    pub(crate) fn export_import_alias(&mut self, k: &str) {
        if let Some(imports) = self.alias_nominal.get(k) {
            for import in imports {
                if let Some(items) = self.imported_items.get_mut(import) {
                    items.exported = true;
                }
            }
        }
    }

    pub(crate) fn insert_imported_items(&mut self, k: ImportId, v: ImportedItems) {
        self.imported_items.insert(k, v);
    }

    pub fn lookup_import_alias(&self, k: &str) -> Option<&[ImportId]> {
        Some(self.alias_nominal.get(k)?)
    }

    pub fn iter_term_item(&self) -> impl Iterator<Item = (TermItemId, &TermItem)> {
        self.term_item.iter()
    }

    pub fn iter_type_item(&self) -> impl Iterator<Item = (TypeItemId, &TypeItem)> {
        self.type_item.iter()
    }
}

impl Relational {
    pub(crate) fn insert_data_relation(&mut self, type_id: TypeItemId, term_id: TermItemId) {
        self.data.push((type_id, term_id));
    }

    pub(crate) fn insert_class_relation(&mut self, type_id: TypeItemId, term_id: TermItemId) {
        self.class.push((type_id, term_id));
    }

    pub(crate) fn insert_chain_relation(&mut self, c_id: InstanceChainId, i_id: InstanceId) {
        self.chain.push((c_id, i_id));
    }

    pub(crate) fn insert_instance_relation(&mut self, i_id: InstanceId, m_id: InstanceMemberId) {
        self.instance.push((i_id, m_id));
    }

    pub fn constructors_of(&self, id: TypeItemId) -> impl Iterator<Item = TermItemId> + '_ {
        self.data.iter().filter_map(
            move |(type_id, term_id)| {
                if id == *type_id {
                    Some(*term_id)
                } else {
                    None
                }
            },
        )
    }

    pub fn class_members_of(&self, id: TypeItemId) -> impl Iterator<Item = TermItemId> + '_ {
        self.class.iter().filter_map(
            move |(type_id, term_id)| {
                if id == *type_id {
                    Some(*term_id)
                } else {
                    None
                }
            },
        )
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum ItemKind {
    DataSignature(DataSignatureId),
    DataEquation(DataEquationId),

    NewtypeSignature(NewtypeSignatureId),
    NewtypeEquation(NewtypeEquationId),

    Constructor(DataConstructorId),

    SynonymSignature(TypeSignatureId),
    SynonymEquation(TypeEquationId),

    ClassSignature(ClassSignatureId),
    ClassDeclaration(ClassDeclarationId),
    ClassMember(ClassMemberId),

    ValueSignature(ValueSignatureId),
    ValueEquation(ValueEquationId),

    ForeignData(ForeignDataId),
    ForeignValue(ForeignValueId),
    Operator(InfixId),
}

#[derive(Debug, PartialEq, Eq)]
pub enum ExistingKind {
    Term(TermItemId),
    Type(TypeItemId),
}

#[derive(Debug, PartialEq, Eq)]
pub enum IndexError {
    DuplicateItem { kind: ItemKind, existing: ExistingKind },
    MismatchedItem { kind: ItemKind, existing: ExistingKind },
    InvalidRole { id: TypeRoleId, existing: Option<TypeItemId> },
    InvalidExport { id: ExportItemId },
    DuplicateExport { id: ExportItemId, existing: ExportItemId },
    DuplicateImport { id: ImportItemId, existing: ImportItemId },
}
