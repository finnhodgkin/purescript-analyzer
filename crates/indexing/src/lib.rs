mod algorithm;

use fxhash::{FxBuildHasher, FxHashMap};
use la_arena::{Arena, ArenaMap, Idx};
use petgraph::{prelude::GraphMap, Directed};
use smol_str::SmolStr;
use syntax::{cst, SyntaxNode, SyntaxNodePtr};

pub type DeclarationId = Idx<cst::Declaration>;

#[derive(Debug, PartialEq, Eq)]
pub struct ClassGroup {
    pub signature: Option<DeclarationId>,
    pub declaration: Option<DeclarationId>,
}

#[derive(Debug, Default)]
pub struct ClassIndex {
    pub by_type: FxHashMap<SmolStr, ClassGroup>,
    pub by_member: FxHashMap<SmolStr, DeclarationId>,
    pub statement_graph: GraphMap<DeclarationId, (), Directed, FxBuildHasher>,
}

#[derive(Debug, Default)]
pub struct InstanceIndex {
    pub by_type: FxHashMap<SmolStr, Vec<DeclarationId>>,
    pub by_name: FxHashMap<SmolStr, DeclarationId>,
    pub instance_graph: GraphMap<DeclarationId, (), Directed, FxBuildHasher>,
    pub statement_graph: GraphMap<DeclarationId, (), Directed, FxBuildHasher>,
}

#[derive(Debug, PartialEq, Eq)]
pub struct ValueGroup {
    pub signature: Option<DeclarationId>,
    pub equations: Vec<DeclarationId>,
}

pub type ValueIndex = FxHashMap<SmolStr, ValueGroup>;

#[derive(Debug, Default)]
pub struct FullIndexingResult {
    arena: Arena<cst::Declaration>,
    pub pointer: ArenaMap<DeclarationId, SyntaxNodePtr>,
    pub class: ClassIndex,
    pub instance: InstanceIndex,
    pub value: ValueIndex,
    pub errors: Vec<IndexingError>,
}

#[derive(Debug, PartialEq, Eq)]
pub enum IndexingError {
    SignatureConflict { existing: DeclarationId, duplicate: DeclarationId },
    SignatureIsLate { equation: DeclarationId, signature: DeclarationId },
    DeclarationConflict { existing: DeclarationId, duplicate: DeclarationId },
}

pub fn index(node: SyntaxNode) -> FullIndexingResult {
    let mut module_map = FullIndexingResult::default();
    algorithm::index_module(&mut module_map, node);
    module_map
}
