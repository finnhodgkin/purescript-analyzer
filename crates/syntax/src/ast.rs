#[macro_use]
mod macros;

mod binder;
mod binding;
mod common;
mod declaration;
mod module;
mod name;

pub use binder::*;
pub use binding::*;
pub use common::*;
pub use declaration::*;
pub use module::*;
pub use name::*;
