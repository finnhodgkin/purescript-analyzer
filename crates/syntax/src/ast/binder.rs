use rowan::ast::AstNode;

use crate::SyntaxToken;

use super::{Name, OneOrMore, QualifiedName};

_create_ast_v!(
    Binder,
    ConstructorBinder(ConstructorBinder),
    LiteralBinder(LiteralBinder),
    NegativeBinder(NegativeBinder),
    ParenthesizedBinder(ParenthesizedBinder),
    TypedBinder(TypedBinder),
    VariableBinder(VariableBinder),
    WildcardBinder(WildcardBinder)
);

impl ConstructorBinder {
    pub fn qualified_name(&self) -> Option<QualifiedName> {
        QualifiedName::cast(self.node.first_child()?)
    }

    pub fn fields(&self) -> Option<OneOrMore<Binder>> {
        OneOrMore::cast(self.node.children().nth(1)?)
    }
}

impl ParenthesizedBinder {
    pub fn binder(&self) -> Option<Binder> {
        Binder::cast(self.node.first_child()?)
    }
}

impl NegativeBinder {
    pub fn minus(&self) -> Option<SyntaxToken> {
        self.node.first_token()
    }

    pub fn literal(&self) -> Option<LiteralBinder> {
        LiteralBinder::cast(self.node.first_child()?)
    }
}

impl VariableBinder {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }
}
