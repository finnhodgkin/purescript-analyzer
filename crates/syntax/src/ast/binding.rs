use rowan::ast::AstNode;

use super::OneOrMore;

_create_ast_v!(Binding, UnconditionalBinding(UnconditionalBinding), GuardedBinding(GuardedBinding));

_create_ast!(GuardedExpression, WhereExpression);

impl UnconditionalBinding {
    pub fn where_expression(&self) -> Option<WhereExpression> {
        WhereExpression::cast(self.node.first_child()?)
    }
}

impl GuardedBinding {
    pub fn guarded_expressions(&self) -> Option<OneOrMore<GuardedExpression>> {
        OneOrMore::cast(self.node.first_child()?)
    }
}

impl GuardedExpression {
    pub fn where_expression(&self) -> Option<WhereExpression> {
        WhereExpression::cast(self.node.last_child()?)
    }
}
