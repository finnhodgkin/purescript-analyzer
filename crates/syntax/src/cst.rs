use rowan::ast::{AstChildren, AstNode};

use crate::{SyntaxKind, SyntaxToken};

#[macro_use]
mod macros;

create_cst_struct!(Module, ModuleHeader, ExportList);

create_cst_enum!(
    ExportItem
        | ExportValue
        | ExportClass
        | ExportType
        | ExportOperator
        | ExportTypeOperator
        | ExportModule
);

create_cst_struct!(ModuleImports, ImportStatement, ImportList, ImportAlias);

create_cst_enum!(
    ImportItem // Comment to fix formatting :P
        | ImportValue
        | ImportClass
        | ImportType
        | ImportOperator
        | ImportTypeOperator
);

create_cst_enum!(TypeItems | TypeItemsAll | TypeItemsList);

create_cst_struct!(ModuleStatements);

create_cst_enum!(
    Declaration
        | ValueSignature
        | ValueEquation
        | InfixDeclaration
        | TypeRoleDeclaration
        | InstanceChain
        | InstanceDeclaration
        | InstanceSignatureStatement
        | InstanceEquationStatement
        | TypeSynonymSignature
        | TypeSynonymEquation
        | ClassSignature
        | ClassDeclaration
        | ClassMemberStatement
        | ForeignImportDataDeclaration
        | ForeignImportValueDeclaration
        | NewtypeSignature
        | NewtypeEquation
        | DataSignature
        | DataEquation
        | DeriveDeclaration
);

associated_declarations!(
    ClassDeclaration where ClassMemberStatement
);

associated_declarations!(
    InstanceDeclaration where InstanceSignatureStatement | InstanceEquationStatement
);

create_cst_struct!(
    TypeRole,
    ClassConstraints,
    ClassHead,
    ClassFunctionalDependencies,
    ClassStatements
);

create_cst_enum!(
    FunctionalDependency | FunctionalDependencyDetermined | FunctionalDependencyDetermines
);

create_cst_struct!(InstanceName, InstanceConstraints, InstanceHead, InstanceStatements);

create_cst_struct!(DataConstructor);

create_cst_enum!(
    Type // Comment to fix formatting :P
        | TypeApplicationChain
        | TypeArrow
        | TypeConstrained
        | TypeConstructor
        | TypeForall
        | TypeHole
        | TypeInteger
        | TypeKinded
        | TypeOperator
        | TypeOperatorChain
        | TypeString
        | TypeVariable
        | TypeVariableBinding
        | TypeWildcard
        | TypeRecord
        | TypeRow
        | TypeParenthesized
);

create_cst_enum!(
    Binder
        | BinderTyped
        | BinderOperatorChain
        | BinderInteger
        | BinderNumber
        | BinderConstructor
        | BinderVariable
        | BinderNamed
        | BinderWildcard
        | BinderString
        | BinderChar
        | BinderTrue
        | BinderFalse
        | BinderArray
        | BinderRecord
        | BinderParenthesized
);

create_cst_struct!(
    FunctionBinders,
    Unconditional,
    WhereExpression,
    Conditionals,
    PatternGuarded,
    PatternGuards
);

create_cst_enum!(PatternGuard | PatternGuardBinder | PatternGuardExpression);

create_cst_enum!(
    Expression
        | ExpressionTyped
        | ExpressionOperatorChain
        | ExpressionInfixChain
        | ExpressionTick
        | ExpressionNegate
        | ExpressionApplicationChain
        | ExpressionTypeArgument
        | ExpressionTermArgument
        | ExpressionIfThenElse
        | ExpressionLetIn
        | ExpressionLambda
        | ExpressionCaseOf
        | ExpressionDo
        | ExpressionAdo
        | ExpressionConstructor
        | ExpressionVariable
        | ExpressionOperatorName
        | ExpressionSection
        | ExpressionHole
        | ExpressionString
        | ExpressionChar
        | ExpressionTrue
        | ExpressionFalse
        | ExpressionInteger
        | ExpressionNumber
        | ExpressionArray
        | ExpressionRecord
        | ExpressionParenthesized
        | ExpressionRecordAccess
        | ExpressionRecordUpdate
);

create_cst_struct!(CaseTrunk, CaseBranches, CaseBranchBinders, CaseBranch);

create_cst_enum!(DoStatement | DoStatementBind | DoStatementLet | DoStatementDiscard);

create_cst_enum!(LetBinding | LetBindingPattern | LetBindingSignature | LetBindingEquation);

create_cst_enum!(RecordItem | RecordField | RecordPun);

create_cst_struct!(RecordUpdates);

create_cst_enum!(RecordUpdate | RecordUpdateLeaf | RecordUpdateBranch);

impl Module {
    pub fn statements(&self) -> Option<ModuleStatements> {
        self.syntax().children().find_map(ModuleStatements::cast)
    }
}

impl ModuleStatements {
    pub fn children(&self) -> impl Iterator<Item = Declaration> {
        self.syntax().children().filter_map(Declaration::cast)
    }
}

impl InstanceChain {
    pub fn instance_declarations(&self) -> AstChildren<InstanceDeclaration> {
        rowan::ast::support::children(self.syntax())
    }
}

impl InstanceDeclaration {
    pub fn instance_head(&self) -> Option<InstanceHead> {
        rowan::ast::support::child(self.syntax())
    }

    pub fn instance_name(&self) -> Option<InstanceName> {
        rowan::ast::support::child(self.syntax())
    }

    pub fn instance_statements(&self) -> Option<InstanceStatements> {
        rowan::ast::support::child(self.syntax())
    }
}

impl InstanceName {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}

impl InstanceHead {
    pub fn type_name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::UPPER)
    }
}

impl InstanceStatements {
    pub fn children(&self) -> AstChildren<InstanceDeclarationStatement> {
        rowan::ast::support::children(self.syntax())
    }
}

impl ValueSignature {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}

impl ValueEquation {
    pub fn name_token(&self) -> Option<SyntaxToken> {
        rowan::ast::support::token(self.syntax(), SyntaxKind::LOWER)
    }
}
