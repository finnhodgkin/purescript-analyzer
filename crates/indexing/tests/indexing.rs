use indexing::{
    Duplicate, ExprItem, FullIndexingResult, Id, IndexingError, TypeGroupId, TypeItem, ValueGroupId,
};
use rowan::ast::AstNode;
use syntax::cst;

fn index<'line>(source: impl AsRef<[&'line str]>) -> (cst::Module, FullIndexingResult) {
    let source = format!("module Main where\n{}", source.as_ref().join("\n"));

    let lexed = lexing::lex(&source);
    let tokens = lexing::layout(&lexed);
    let (node, _) = parsing::parse(&lexed, &tokens);

    let module = cst::Module::cast(node).unwrap();
    let index = indexing::index(&module);

    (module, index)
}

#[test]
fn valid_module() {
    let (module, index) = index(&[
        "life :: Int",
        "life = 42",
        "data Boolean :: Type",
        "data Boolean = True | False",
        "newtype Id :: Type -> Type",
        "newtype Id a = Id a",
        "type Fn :: Type -> Type -> Type",
        "type Fn a b = a -> b",
        "class Eq :: Type -> Constraint",
        "class Eq a where",
        "  eq :: a -> a -> Boolean",
        "instance eqInt :: Eq Int where",
        "  eq = eqIntImpl",
    ]);

    assert!(index.errors.is_empty());

    assert!(matches!(index.nominal.lookup_type_item("Boolean"), Some((_, TypeItem::Data(_)))));
    assert!(matches!(index.nominal.lookup_type_item("Id"), Some((_, TypeItem::Newtype(_)))));
    assert!(matches!(index.nominal.lookup_type_item("Fn"), Some((_, TypeItem::Synonym(_)))));
    assert!(matches!(index.nominal.lookup_type_item("Eq"), Some((_, TypeItem::Class(_)))));

    if let Some((_, ExprItem::Value(ValueGroupId { signature: Some(id), .. }))) =
        index.nominal.lookup_expr_item("life")
    {
        let ptr = index.source_map.declaration_ptr(*id).unwrap();
        let node = ptr.to_node(module.syntax());

        assert_eq!(node.syntax().text(), "\nlife :: Int");
        assert_eq!(index.source_map.declaration_id(ptr), Some(*id));
    } else {
        unreachable!()
    }

    if let Some((_, ExprItem::Method(id))) = index.nominal.lookup_expr_item("eq") {
        let ptr = index.source_map.class_member_ptr(*id).unwrap();
        let node = ptr.to_node(module.syntax());

        assert_eq!(node.syntax().text(), "\n  eq :: a -> a -> Boolean");
        assert_eq!(index.source_map.class_member_id(ptr), Some(*id));
    } else {
        unreachable!()
    }

    if let Some((_, ExprItem::Instance(id))) = index.nominal.lookup_expr_item("eqInt") {
        let ptr = index.source_map.instance_ptr(*id).unwrap();
        let node = ptr.to_node(module.syntax());

        assert_eq!(node.syntax().text(), "\ninstance eqInt :: Eq Int where\n  eq = eqIntImpl");
        assert_eq!(index.source_map.instance_id(ptr), Some(*id));
    } else {
        unreachable!()
    }

    if let Some((_, ExprItem::Constructor(id))) = index.nominal.lookup_expr_item("True") {
        let ptr = index.source_map.constructor_ptr(*id).unwrap();
        let node = ptr.to_node(module.syntax());

        assert_eq!(node.syntax().text(), " True");
        assert_eq!(index.source_map.constructor_id(ptr), Some(*id));
    } else {
        unreachable!()
    }

    if let Some((_, ExprItem::Constructor(id))) = index.nominal.lookup_expr_item("Id") {
        if let Some(id) = index.relational.of_constructor(*id) {
            assert!(matches!(index.nominal.index_type_item(id), Some(TypeItem::Newtype(_))));
        } else {
            unreachable!()
        }
    } else {
        unreachable!()
    }

    if let Some((_, ExprItem::Method(id))) = index.nominal.lookup_expr_item("eq") {
        if let Some(id) = index.relational.of_method(*id) {
            assert!(matches!(index.nominal.index_type_item(id), Some(TypeItem::Class(_))));
        } else {
            unreachable!()
        }
    } else {
        unreachable!()
    }

    let (id, item) = index.nominal.lookup_expr_item("life").unwrap();
    assert_eq!(index.nominal.index_expr_item(id), Some(item));

    if let Some((id, _)) = index.nominal.lookup_type_item("Boolean") {
        let constructors: Vec<_> = index.relational.constructors_of(id).collect();
        if let &[t, f] = &constructors[..] {
            let t_ptr = index.source_map.constructor_ptr(t).unwrap();
            let t_node = t_ptr.to_node(module.syntax());

            assert_eq!(t_node.syntax().text(), " True");
            assert_eq!(index.source_map.constructor_id(t_ptr), Some(t));

            let f_ptr = index.source_map.constructor_ptr(f).unwrap();
            let f_node = f_ptr.to_node(module.syntax());

            assert_eq!(f_node.syntax().text(), " False");
            assert_eq!(index.source_map.constructor_id(f_ptr), Some(f));
        } else {
            unreachable!()
        }
    } else {
        unreachable!()
    }

    if let Some((id, _)) = index.nominal.lookup_type_item("Id") {
        let constructors: Vec<_> = index.relational.constructors_of(id).collect();
        if let &[id] = &constructors[..] {
            let ptr = index.source_map.constructor_ptr(id).unwrap();
            let node = ptr.to_node(module.syntax());

            assert_eq!(node.syntax().text(), " Id a");
            assert_eq!(index.source_map.constructor_id(ptr), Some(id));
        } else {
            unreachable!()
        }
    } else {
        unreachable!()
    }

    if let Some((id, _)) = index.nominal.lookup_type_item("Eq") {
        let methods: Vec<_> = index.relational.methods_of(id).collect();
        if let &[id] = &methods[..] {
            let ptr = index.source_map.class_member_ptr(id).unwrap();
            let node = ptr.to_node(module.syntax());

            assert_eq!(node.syntax().text(), "\n  eq :: a -> a -> Boolean");
            assert_eq!(index.source_map.class_member_id(ptr), Some(id));
        } else {
            unreachable!()
        }
    } else {
        unreachable!()
    }

    if let Some((_, TypeItem::Synonym(TypeGroupId { declaration: Some(id), .. }))) =
        index.nominal.lookup_type_item("Fn")
    {
        let ptr = index.source_map.declaration_ptr(*id).unwrap();
        let node = ptr.to_node(module.syntax());

        assert_eq!(node.syntax().text(), "\ntype Fn a b = a -> b");
        assert_eq!(index.source_map.declaration_id(ptr), Some(*id));
    } else {
        unreachable!()
    }
}

#[test]
fn late_signature_error() {
    let (_, index) = index(&[
        "life = 42",
        "life :: Int",
        "data Boolean = True | False",
        "data Boolean :: Type",
        "newtype Id a = Id a",
        "newtype Id :: Type -> Type",
        "type Fn a b = a -> b",
        "type Fn :: Type -> Type -> Type",
        "class Eq a",
        "class Eq :: Type -> Constraint",
    ]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::LateSignature {
                declaration: Id::from_raw(0),
                signature: Id::from_raw(1)
            },
            IndexingError::LateSignature {
                declaration: Id::from_raw(2),
                signature: Id::from_raw(3)
            },
            IndexingError::LateSignature {
                declaration: Id::from_raw(4),
                signature: Id::from_raw(5)
            },
            IndexingError::LateSignature {
                declaration: Id::from_raw(6),
                signature: Id::from_raw(7)
            },
            IndexingError::LateSignature {
                declaration: Id::from_raw(8),
                signature: Id::from_raw(9)
            },
        ]
    );
}

#[test]
fn duplicate_expr_item() {
    let (_, index) = index(&[
        "life :: Int",
        "life = 42",
        "instance life :: Life",
        "class Life where",
        "  life :: Int",
        "life :: Int",
        "data IdData a = Id a",
        "newtype IdNewtype a = Id a",
    ]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Instance(Id::from_raw(0)),
            },
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::ClassMember(Id::from_raw(0)),
            },
            IndexingError::LateSignature {
                declaration: Id::from_raw(1),
                signature: Id::from_raw(4)
            },
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(4)),
            },
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(1),
                duplicate: Duplicate::Constructor(Id::from_raw(1)),
            },
        ]
    );
}

#[test]
fn value_is_duplicate() {
    let (_, index) = index([
        "class Eq a where",
        "  eq :: a -> a -> Boolean",
        "eq :: Int -> Int -> Boolean",
        "eq = eqIntImpl",
    ]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateExprItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(2)),
            }
        ]
    );
}

#[test]
fn duplicate_data() {
    let (_, index) =
        index(&["data Id0 :: Type", "data Id0 :: Type", "data Id1 = Id2", "data Id1 = Id3"]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(1),
                duplicate: Duplicate::Declaration(Id::from_raw(3)),
            },
        ]
    );
}

#[test]
fn duplicate_newtype() {
    let (_, index) = index(&[
        "newtype Id0 :: Type",
        "newtype Id0 :: Type",
        "newtype Id1 = Id2 Int",
        "newtype Id1 = Id3 Int",
    ]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(1),
                duplicate: Duplicate::Declaration(Id::from_raw(3)),
            },
        ]
    );
}

#[test]
fn duplicate_class() {
    let (_, index) = index(&["class Id0 :: Type", "class Id0 :: Type", "class Id1", "class Id1"]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(1),
                duplicate: Duplicate::Declaration(Id::from_raw(3)),
            },
        ]
    );
}

#[test]
fn duplicate_synonym() {
    let (_, index) =
        index(&["type Id0 :: Type", "type Id0 :: Type", "type Id1 = Id0", "type Id1 = Id0"]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(1),
                duplicate: Duplicate::Declaration(Id::from_raw(3)),
            },
        ]
    );
}

#[test]
fn duplicate_type_item() {
    let (_, index) = index(&[
        "data Id0 :: Type",
        "data Id0 = Id10 Int",
        "newtype Id0 :: Type",
        "newtype Id0 = Id11 Int",
        "class Id0 :: Type",
        "class Id0",
        "type Id0 :: Type",
        "type Id0 = Id0",
    ]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(2)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(3)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(4)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(5)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(6)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(7)),
            },
        ]
    );
}

#[test]
fn data_is_duplicate() {
    let (_, index) = index(&["class Eq", "data Eq :: Type", "data Eq"]);
    assert_eq!(
        &index.errors,
        &[
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(1)),
            },
            IndexingError::DuplicateTypeItem {
                item_id: Id::from_raw(0),
                duplicate: Duplicate::Declaration(Id::from_raw(2)),
            }
        ]
    );
}
