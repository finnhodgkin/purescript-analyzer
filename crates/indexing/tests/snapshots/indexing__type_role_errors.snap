---
source: crates/indexing/tests/indexing.rs
expression: errors
snapshot_kind: text
---
[
    InvalidRole {
        role: Id(
            1,
        ),
    },
    EmptyRole {
        role: Id(
            2,
        ),
    },
    EmptyRole {
        role: Id(
            4,
        ),
    },
    EmptyRole {
        role: Id(
            5,
        ),
    },
    DuplicateDeclaration {
        declaration: Id(
            4,
        ),
        duplicate: Id(
            5,
        ),
    },
]
